Userland/Games/Chess/PromotionDialog.cpp

```diff
@@ -9,31 +9,51 @@
 #include <LibGUI/Button.h>
 #include <LibGUI/Frame.h>
 
-PromotionDialog::PromotionDialog(ChessWidget& chess_widget)
-    : Dialog(chess_widget.window())
+namespace Chess {
+
+ErrorOr<NonnullRefPtr<PromotionDialog>> PromotionDialog::try_create(ChessWidget& chess_widget)
+{
+    auto promotion_widget = TRY(Chess::PromotionWidget::try_create());
+    auto promotion_dialog = TRY(adopt_nonnull_ref_or_enomem(new (nothrow) PromotionDialog(move(promotion_widget), chess_widget)));
+    return promotion_dialog;
+}
+
+PromotionDialog::PromotionDialog(NonnullRefPtr<Chess::PromotionWidget> promotion_widget, ChessWidget& chess_widget)
+    : GUI::Dialog(chess_widget.window())
     , m_selected_piece(Chess::Type::None)
 {
-    set_title("Choose piece to promote to");
+    set_title("Choose piece to promote to...");
     set_icon(chess_widget.window()->icon());
-    resize(70 * 4, 70);
-
-    auto main_widget = set_main_widget<GUI::Frame>();
-    main_widget->set_frame_style(Gfx::FrameStyle::SunkenContainer);
-    main_widget->set_fill_with_background_color(true);
-    main_widget->set_layout<GUI::HorizontalBoxLayout>();
-
-    for (auto const& type : { Chess::Type::Queen, Chess::Type::Knight, Chess::Type::Rook, Chess::Type::Bishop }) {
-        auto& button = main_widget->add<GUI::Button>();
-        button.set_fixed_height(70);
-        button.set_icon(chess_widget.get_piece_graphic({ chess_widget.board().turn(), type }));
-        button.on_click = [this, type](auto) {
-            m_selected_piece = type;
-            done(ExecResult::OK);
-        };
-    }
+    set_resizable(false);
+
+    set_main_widget(promotion_widget);
+
+    auto queen_button = promotion_widget->find_descendant_of_type_named<GUI::Button>("queen_button");
```

Comment:
Something like this could also work here:

```c++
auto initialize_promotion_button = [&](StringView button_name, Chess::Type piece) {
    auto button = promotion_widget->find_descendant_of_type_named<GUI::Button>(button_name);
    button->set_icon(chess_widget.get_piece_graphic({ chess_widget.board().turn(), piece }));
    button->on_click = [this, piece](auto) {
        m_selected_piece = piece;
        done(ExecResult::OK);
    };
};

initialize_promotion_button("queen_button"sv, Type::Queen);
initialize_promotion_button("knight_button"sv, Type::Knight);
initialize_promotion_button("rook_button"sv, Type::Rook);
initialize_promotion_button("bishop_button"sv, Type::Bishop);
```

Reply:
Yep, that works and it's much cleaner, thank you!